### PR TITLE
feat: Implement preference for exit on close behavior, defaulting to true on Linux

### LINKED ISSUE #471: Add option to disable running in system tray
### Is your enhancement related to a problem? Please describe

App keeps running minimized and in system tray after closing it. This might not be desired by some users, personally I would much prefer it if clicking the close button actually quit the application. Additionally, some Linux desktop environments (eg. GNOME) ship without tray icon support by default. This means some users might not even be aware that the application keeps running after "closing" it, which is poor UX.

### Describe the solution you'd like

Add an option that disables the tray icon and quits the application when the window is closed.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dylanmtaylor**
Signed-off-by: Dylan M. Taylor <dylan@dylanmtaylor.com>

### What does this PR do?

This pull request alters the behavior of the application when running on Linux. When the close button is clicked, the application will close like a normal application would, instead of attempting to run  in the background still or minimize to tray.

The rationale for this change is as follows:

* On GNOME systems, such as Fedora Workstation or RHEL, by default there is no system tray. 
  * The workaround for this is ugly - both from a UX perspective as well as cosmetically -- having a menu bar and expecting the user to do File > Quit to close out of the application
  * I made another PR, #668 to hide the menu bar, and the client looks MUCH better IMO, but now a user still has to use the menu bar to close the app normally, and they have to hold Alt to get it to appear.
* Linux is a very different beast with podman than Mac or Windows. While it's super inconvenient to have the podman machine VM unexpectedly close on Mac/Windows, this is simply not a problem on Linux. The GUI for podman desktop simply attaches to a socket, so closing the app doesn't necessarily terminate podman containers. Users will _expect_ that the close button will close the application, and not having it do so will be a bad user experience.

### What issues does this PR fix or reference?

Issue #663, and this goes with PR #668.

Closes #471

### How to test this PR?

Open the application on Linux and then click the X to close it. Validate that it did, in fact, close.

This has been tested by me to work, but it isn't exactly a complicated change.

### CODE REVIEW

**Comment by dylanmtaylor** [comment]
Another thing that might work -- under settings, have Preferences > Minimize to Tray on Close. Default to no on Linux, yes on other platforms. Let this setting control the close behavior.

**Comment by benoitf** [comment]
hello @dylanmtaylor 
We've changed default behaviour on Linux quite often so yes I think it would be better if we have the ability in preferences to toggle the default action on Linux for closing.

**Comment by dylanmtaylor** [comment]
> hello @dylanmtaylor We've changed default behaviour on Linux quite often so yes I think it would be better if we have the ability in preferences to toggle the default action on Linux for closing.

Could you help me figure out how to make a PR to implement that? I'm willing to help out a bit in my free time. I imagine it'll need a UI element in the front end, a value in the preferences store, and some logic to retrieve the value when the close button is clicked, to determine close vs hide.

**Comment by benoitf** [comment]
sure, I think I would need to make first, some changes in the configuration system to be initialized early as for now everything is initialized once the window is created so it'll likely be too late

**Comment by benoitf** [comment]
How to get for now the configurationRegistry
https://github.com/containers/podman-desktop/blob/main/packages/main/src/plugin/index.ts#L211-L212

How to register a key:
https://github.com/containers/podman-desktop/blob/f41b578a4b3c06d52a42ee8ba5c5fa53673e0b5d/packages/main/src/plugin/autostart-engine.ts#L25-L40

How to retrieve the value for the key
https://github.com/containers/podman-desktop/blob/f41b578a4b3c06d52a42ee8ba5c5fa53673e0b5d/packages/main/src/plugin/autostart-engine.ts#L46

For now you may just miss a reference to the configurationRegistry (as it's created/initialized at a later scope)

**Comment by benoitf** [comment]
also here is the issue to have that behaviour: https://github.com/containers/podman-desktop/issues/471

**Comment by dylanmtaylor** [comment]
> How to get for now the configurationRegistry https://github.com/containers/podman-desktop/blob/main/packages/main/src/plugin/index.ts#L211-L212
> 
> How to register a key:
> 
> https://github.com/containers/podman-desktop/blob/f41b578a4b3c06d52a42ee8ba5c5fa53673e0b5d/packages/main/src/plugin/autostart-engine.ts#L25-L40
> 
> How to retrieve the value for the key
> 
> https://github.com/containers/podman-desktop/blob/f41b578a4b3c06d52a42ee8ba5c5fa53673e0b5d/packages/main/src/plugin/autostart-engine.ts#L46
> 
> For now you may just miss a reference to the configurationRegistry (as it's created/initialized at a later scope)

Yeah, I'm struggling to see how to wire up retrieving a configuration property from the mainWindow.ts file

**Comment by dylanmtaylor** [comment]
@benoitf if you have some time, could you take a look at the work in progress commit (I'll rebase and force push this cleanly), and let me know what I need to do to make it actually function. As you suggested, getting the preferences in this function is going to be hard.

**Comment by dylanmtaylor** [comment]
I'm trying to learn a bit here too, my JavaScript isn't the strongest. I think I need to find a way to pass the configurationRegistry to the mainWindow.ts file once that is ready. I see it all initialized in the index.ts.

```
app.whenReady().then(async () => {
  const animatedTray = new AnimatedTray();
  tray = new Tray(animatedTray.getDefaultImage());
  animatedTray.setTray(tray);
  const trayMenu = new TrayMenu(tray, animatedTray);
  // start extensions
  const pluginSystem = new PluginSystem(trayMenu);
  const extensionLoader = await pluginSystem.initExtensions();

  const configurationRegistry = extensionLoader.getConfigurationRegistry();

  // configure automatic startup
  const automaticStartup = new StartupInstall(configurationRegistry);
  await automaticStartup.configure();
});
```

I don't think the value needs to be ready per-se at the time of construction as what happens when it's closed isn't a needed property until the close button is clicked.

**Review by benoitf — COMMENTED** [review]
I don't see where init method or the constructor is called ?

Also in close, the configuration registry might be undefined so it should use ?. property access

File: packages/main/src/plugin/close-behavior.ts
Comment (benoitf) [inline packages/main/src/plugin/close-behavior.ts]: Maybe we should register the property only for Linux users

Comment (dylanmtaylor) [reply packages/main/src/plugin/close-behavior.ts]: I think allowing customization across OSes would be valuable. The only thing I think should be different is the default behavior. That way it's more consistent.

**Comment by dylanmtaylor** [comment]
> so it should use ?. property access

I added the object construction and init() call. Can you clarify how I'd use this property access? I'm not a very seasoned JS developer.

File: packages/main/src/mainWindow.ts
Comment (benoitf) [inline packages/main/src/mainWindow.ts]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

It's because configuration registry may not be defined there

Comment (benoitf) [reply packages/main/src/mainWindow.ts]: also if configuration is not defined we shouldn't try to get the property as well

Comment (benoitf) [reply packages/main/src/mainWindow.ts]: You should probably declare a variable `configurationRegistry` at an upper scope and remove the const in the initialization after ready state is reached

Comment (dylanmtaylor) [reply packages/main/src/mainWindow.ts]: I am trying to declare it global, but it seems like I keep getting the value undefined in the actual onClose call.

**Comment by dylanmtaylor** [comment]
Oh, so this is interesting. With the latest code, it *does* seem to work when minimize to close is checked, returning true, but false returns undefined. Odd.

```
  browserWindow.on('close', e => {
    // console.log(global.configurationRegistry);
    const closeBehaviorConfiguration = global.configurationRegistry.getConfiguration('preferences');
    console.log(closeBehaviorConfiguration);
    const minimizeonclose = closeBehaviorConfiguration.get<boolean>('MinimizeToTrayOnClose');
    console.log("Minimize on close: " + minimizeonclose);
    if (minimizeonclose) {
      e.preventDefault();
      browserWindow.hide();
      if (isMac) {
        app.dock.hide();
      }
    } 
  });
```

```
7:33:25 PM [main] ConfigurationImpl {
  scope: 'DEFAULT',
  configurationValues: Map(2) {
    'DEFAULT' => {
      'telemetry.check': true,
      'telemetry.enabled': true,
      'preferences.MinimizeToTrayOnClose': true,
      'preferences.engine.autostart': true,
      'terminal.integrated.fontSize': 10,
      'terminal.integrated.lineHeight': 1
    },
    'container-connection:/Users/taylord/.local/share/containers/podman/machine/podman-machine-default/podman.sock' => {
      'podman.machine.cpus': 1,
      'podman.machine.memory': 2048,
      'podman.machine.diskSize': 100
    }
  },
  globalSection: 'preferences'
}

7:33:25 PM [main] Minimize on close: true
```

But when it is false, it shows undefined.

```
7:33:59 PM [main] ConfigurationImpl {
  scope: 'DEFAULT',
  configurationValues: Map(2) {
    'DEFAULT' => {
      'telemetry.check': true,
      'telemetry.enabled': true,
      'preferences.MinimizeToTrayOnClose': false,
      'preferences.engine.autostart': true,
      'terminal.integrated.fontSize': 10,
      'terminal.integrated.lineHeight': 1
    },
    'container-connection:/Users/taylord/.local/share/containers/podman/machine/podman-machine-default/podman.sock' => {
      'podman.machine.cpus': 1,
      'podman.machine.memory': 2048,
      'podman.machine.diskSize': 100
    }
  },
  globalSection: 'preferences'
}

7:33:59 PM [main] Minimize on close: undefined
```

If I do this, it has the correct value:

```
  browserWindow.on('close', e => {
    // console.log(global.configurationRegistry);
    const closeBehaviorConfiguration = global.configurationRegistry.getConfiguration('preferences');
    console.log(closeBehaviorConfiguration);
    const minimizeonclose: boolean = closeBehaviorConfiguration.get<boolean>('MinimizeToTrayOnClose') == true;
    console.log("Minimize on close: " + minimizeonclose);
    if (minimizeonclose) {
      e.preventDefault();
      browserWindow.hide();
      if (isMac) {
        app.dock.hide();
      }
    } 
  }); 
```

```
7:35:04 PM [main] ConfigurationImpl {
  scope: 'DEFAULT',
  configurationValues: Map(2) {
    'DEFAULT' => {
      'telemetry.check': true,
      'preferences.MinimizeToTrayOnClose': false,
      'telemetry.enabled': true,
      'preferences.engine.autostart': true,
      'terminal.integrated.fontSize': 10,
      'terminal.integrated.lineHeight': 1
    },
    'container-connection:/Users/taylord/.local/share/containers/podman/machine/podman-machine-default/podman.sock' => {
      'podman.machine.cpus': 1,
      'podman.machine.memory': 2048,
      'podman.machine.diskSize': 100
    }
  },
  globalSection: 'preferences'
}

7:35:04 PM [main] Minimize on close: false
```

**Comment by dylanmtaylor** [comment]
Tested to function correctly in both true and false cases on Mac OS and Linux. :) This PR is ready for review. Honestly, I must say, I'm extremely happy with how this works now, and I'm hoping it gets accepted. I've tested it on GNOME and KDE, and it behaves perfectly. The force pushes were me changing some things to make the linter happy. I can now report that all the tests are passing.

File: packages/main/src/mainWindow.ts
Comment (dylanmtaylor) [reply packages/main/src/mainWindow.ts]: It turns out, this property is always ready by the time the window is open. Not sure if this is needed?

File: packages/main/src/index.ts
Comment (benoitf) [inline packages/main/src/index.ts]: you can declare it like the tray object
```suggestion
let configurationRegistry: ConfigurationRegistry;
```

File: packages/main/src/index.ts
Comment (benoitf) [inline packages/main/src/index.ts]:
```
configurationRegistry = extensionLoader.getConfigurationRegistry();
```

File: packages/main/src/index.ts
Comment (benoitf) [inline packages/main/src/index.ts]: you can revert as before

File: packages/main/src/mainWindow.ts
Comment (benoitf) [inline packages/main/src/mainWindow.ts]: I think that for this PR, we shouldn't address the menu bar item.
It's all about having a way to hide or exit on close

File: packages/main/src/mainWindow.ts
Comment (benoitf) [inline packages/main/src/mainWindow.ts]:
```suggestion
    const closeBehaviorConfiguration = configurationRegistry?.getConfiguration('preferences');
```

File: packages/main/src/mainWindow.ts
Comment (benoitf) [inline packages/main/src/mainWindow.ts]:
```suggestion
    if (closeBehaviorConfiguration) {
```

File: packages/main/src/mainWindow.ts
Comment (benoitf) [inline packages/main/src/mainWindow.ts]: you can delete this log instruction when you're ready with the PR

File: packages/main/src/plugin/close-behavior.ts
Comment (benoitf) [inline packages/main/src/plugin/close-behavior.ts]: Maybe the property should be the opposite: `exitOnClose`
WDYT ?

File: packages/main/src/mainWindow.ts
Comment (dylanmtaylor) [reply packages/main/src/mainWindow.ts]: Sure. I thought these we tightly coupled, so I figured we could address both in one PR, but that's fine. I'll remove this change.

File: packages/main/src/plugin/close-behavior.ts
Comment (dylanmtaylor) [reply packages/main/src/plugin/close-behavior.ts]: I'm fine with it either way. I'll change it.

**Comment by dylanmtaylor** [comment]
@benoitf It seems that without it explicitly declared as a global variable, that is, using just let, it fails to run correctly. 

image

I'm going to undo that last commit to apply the changes you suggested, but I'll keep the line with the undefined to a simple

     if (closeBehaviorConfiguration) {

File: packages/main/src/index.ts
Comment (dylanmtaylor) [reply packages/main/src/index.ts]: This didn't work.

File: packages/main/src/index.ts
Comment (dylanmtaylor) [reply packages/main/src/index.ts]: This also didn't work.

File: packages/main/src/index.ts
Comment (benoitf) [reply packages/main/src/index.ts]:
```
let configurationRegistry: ConfigurationRegistry | undefined;
```
It should work as for tray variable
it's also initialized in the same block that were you want to assign configurationRegistry

Comment (benoitf) [inline packages/main/src/index.ts]: I still think it should be like for tray, no need to use global
else it's looking very good

Comment (dylanmtaylor) [reply packages/main/src/index.ts]: I tried without global, and made every change you suggested. It couldn't find it in the main window and gave this error.
image

Comment (dylanmtaylor) [reply packages/main/src/index.ts]: Would you be fine merging as-is and leaving this for future troubleshooting/optimization by someone who knows what they're doing more than I do?

Comment (benoitf) [reply packages/main/src/index.ts]: I'll provide you more in-depth review tomorrow (or provide a patch to this PR)

Comment (benoitf) [reply packages/main/src/index.ts]: @dylanmtaylor I've provided a PR to your branch
https://github.com/dylanmtaylor/podman-desktop/pull/4

And now I'm understanding why you had the undefined error. I thought you were using the object in the same file.

Comment (dylanmtaylor) [reply packages/main/src/index.ts]: I've merged that in

**Review by benoitf — APPROVED** [review]
Thanks @dylanmtaylor for this contribution 🎉 

Tried on macOS and worked as expected

**Comment by dylanmtaylor** [comment]
> Thanks @dylanmtaylor for this contribution 🎉
> 
> Tried on macOS and worked as expected

Thank you for the help, I'd love to see the auto hide menubar go in as well once this is merged. That's a major aesthetic win, in my opinion. :)

**Comment by dylanmtaylor** [comment]
@benoitf All tests have passed, but I appear to lack the ability to click merge. Would you mind doing that for me? I'd suggest not squashing as to maintain the history since we both worked on this. Understandably, it says I don't have write access. Thanks!

**Comment by benoitf** [comment]
@dylanmtaylor I'll probably squash on merge so it'll show your name
people can still get original commits coming back to this PR.

**Comment by dylanmtaylor** [comment]
> @dylanmtaylor I'll probably squash on merge so it'll show your name people can still get original commits coming back to this PR.

That's fine, I'm just happy to have my change merged 🙂

**Pull request merged**
